tests/database/test_journal_db.py

```diff
@@ -47,30 +50,113 @@ def test_revert_clears_reverted_journal_entries(journal_db):
 
     assert journal_db.get(b'1') == b'test-a'
 
-    snapshot_a = journal_db.snapshot()
+    changeset_a = journal_db.record()
 
     journal_db.set(b'1', b'test-b')
     journal_db.delete(b'1')
     journal_db.set(b'1', b'test-c')
 
     assert journal_db.get(b'1') == b'test-c'
 
-    snapshot_b = journal_db.snapshot()
+    changeset_b = journal_db.record()
 
     journal_db.set(b'1', b'test-d')
     journal_db.delete(b'1')
     journal_db.set(b'1', b'test-e')
 
     assert journal_db.get(b'1') == b'test-e'
 
-    journal_db.revert(snapshot_b)
+    journal_db.forget(changeset_b)
 
     assert journal_db.get(b'1') == b'test-c'
 
     journal_db.delete(b'1')
 
     assert journal_db.exists(b'1') is False
 
-    journal_db.revert(snapshot_a)
+    journal_db.forget(changeset_a)
+
+    assert journal_db.get(b'1') == b'test-a'
+
+def test_revert_removes_journal_entries(journal_db):
+
+    changeset_a = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 2
+
+    changeset_b = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 3
+
+    # Forget *latest* changeset and prove it's the only one removed
+    journal_db.forget(changeset_b)
+    assert len(journal_db.journal.journal_data) == 2
+
+    changeset_b2 = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 3
+
+    changeset_c = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 4
 
+    changeset_d = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 5
+
+    # Forget everything from b2 (inclusive) and what follows
+    journal_db.forget(changeset_b2)
+    assert len(journal_db.journal.journal_data) == 2
+    assert journal_db.journal.has_checkpoint(changeset_b2) is False
+
+
+def test_commit_merges_changeset_into_previous(journal_db):
+
+    changeset = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 2
+
+    journal_db.set(b'1', b'test-a')
     assert journal_db.get(b'1') == b'test-a'
+
+    journal_db.commit(changeset)
+
+    assert len(journal_db.journal.journal_data) == 1
+    assert journal_db.journal.has_checkpoint(changeset) is False
+
+def test_committing_middle_changeset_merges_in_subsequent_changesets(journal_db):
+
+    journal_db.set(b'1', b'test-a')
+    changeset_a = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 2
+
+    journal_db.set(b'1', b'test-b')
+    changeset_b = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 3
+
+    journal_db.set(b'1', b'test-c')
+    changeset_c = journal_db.record()
+    assert len(journal_db.journal.journal_data) == 4
+
+    # TODO: Clarify
```

Comment:
I added this test to prove I understood how the system is supposed to handle the case that we commit `C1` where `C2` and `C3` also exist. In fact, committing `C1` merges in `C2` and `C3` but I'm not sure I understand the **why** of that. What would be wrong if committing `C1` would not make assumptions about what happens to `C2` or `C3` just yet?

Reply:
This is quite surprising to me. If I start a transaction `Turkey` and inside it start transaction `Duck`, but never commit transaction `Duck`, then I expect that committing transaction `Turkey` should either:

1) Fail _<- my preference_
2) Ignore the uncommitted Duck transactions